cmd/kubelet/app/server.go

```diff
@@ -412,13 +431,249 @@ func run(s *options.KubeletServer, kcfg *KubeletConfig) (err error) {
+		return fmt.Errorf("unable to load auth info in bootstrap kubeconfig: %v", err)
+	}
+
+	authInfo.ClientCertificate, authInfo.ClientKey, err = getClientCertAndKey(s, authInfo.ClientCertificate, authInfo.ClientKey)
+	if err != nil {
+		return fmt.Errorf("unable to get cert from API server: %v", err)
+	}
+
+	// Marshal and write the kubeconfig to disk.
+	data, err := json.Marshal(kcfg)
+	if err != nil {
+		return fmt.Errorf("unable to marshal the kubeconfig: %v", err)
+	}
+
+	if err := ioutil.WriteFile(s.KubeConfig.Value(), data, 0644); err != nil {
+		return fmt.Errorf("unable to write the kubeconfig file at %q: %v", s.KubeConfig.Value(), err)
+	}
+
+	return nil
+}
+
+// getCurrentContextAuthInfo returns the AuthInfo object that's referenced
+// by the current context.
+// If current context or auth info name is empty, then it will return an error.
+// If AuthInfo is empty, a new auth info object will be created.
+func getCurrentContextAuthInfo(config *clientcmdapi.Config) (*clientcmdapi.AuthInfo, error) {
+	ctx, ok := config.Contexts[config.CurrentContext]
+	if !ok {
+		return nil, fmt.Errorf("unable to find current context %q", config.CurrentContext)
+	}
+
+	if ctx.AuthInfo == "" {
+		return nil, fmt.Errorf("unable to find the name of the authInfo in current context %q", config.CurrentContext)
+	}
+
+	if _, ok := config.AuthInfos[ctx.AuthInfo]; !ok {
+		config.AuthInfos[ctx.AuthInfo] = clientcmdapi.NewAuthInfo()
+	}
+	return config.AuthInfos[ctx.AuthInfo], nil
+}
+
+// TODO(yifan): More detailed check on the cert / key content.
+// CheckTLSInfo returns true if the authInfo contains client certificate data for client key data.
+// Or if the client certificate or key file exists.
+func containsSufficientTLSInfo(authInfo *clientcmdapi.AuthInfo) bool {
+	// We use '||' so that we won't override existing data in case of wrong setup.
+	if len(authInfo.ClientCertificateData) > 0 || len(authInfo.ClientKeyData) > 0 {
+		return true
+	}
+
+	if crypto.FoundCertOrKey(authInfo.ClientCertificate, authInfo.ClientKey) {
+		return true
+	}
+
+	return false
+}
+
+// getClientCertAndKey will:
+// (1) Create a restful client for doing the certificate signing request.
+// (2) Read existing key data from existingKeyPath if possible.
+// (3) Pass 'requestClientCertificate()' the CSR client, existing key data, and node name to
+//     request for client certificate from the API server.
+// (4) Once (3) succeeds, dump the certificate and key data to the given paths.
+// On failure, the the certificate and key file will be cleaned up.
+// If the existingCertPath or existingKeyPath is empty, then the function will use the default path, respectively:
+// /var/run/kubernetes/kubelet-client.crt, /var/run/kubernetes/kubelet-client.key.
+func getClientCertAndKey(s *options.KubeletServer, existingCertPath, existingKeyPath string) (certPath, keyPath string, err error) {
+	// (1).
+	clientConfig, err := kubeconfigClientConfig(s.BootstrapKubeconfig, s.APIServerList)
```

Comment:
load a restclient.Config from the bootstrap kubeconfig, don't include the api server list